### ISSUE TITLE
Migrate to Jakarta EE APIs

### ISSUE THREAD

**Issue by thetric** [rest-assured/rest-assured#1510]
Hi there,

is there a roadmap to migrate rest-assured to Jakarta EE APIs? For example the Spring Framework is planning to migrate in the next year:

> Spring Framework 6 and Spring Boot 3 are planned towards a high-end baseline for their general availability in Q4 2022:
>
> * Java 17+ (from Java 8-17 in the Spring Framework 5.3.x line)
> * Jakarta EE 9+ (from Java EE 7-8 in the Spring Framework 5.3.x line)
>
> - https://spring.io/blog/2021/09/02/a-java-17-and-jakarta-ee-9-baseline-for-spring-framework-6

More information from the Eclipse Foundation: https://www.eclipse.org/community/eclipse_newsletter/2020/november/1.php

**Comment by holgerstolzenberg** [comment]
As the issue has been opened some days ago now - it would be great to hear something about the teams plans upon this topic.

We are using restassured everywhere on our platform currently in conjunction with Spring Boot 2. We are really looking forward on using Spring Boot 3, especially because of JDK-17 and the native stuff coming.

We are very pleased with `restassured` so far and would really like to stick with it. 


@wilkinsona What consequences will we face if we upgrade to Spring Boot 3 without having `restassured` supporting the given APIs?

**Comment by wilkinsona** [comment]
That's really a question for the REST Assured team if you want a definitive answer. As far as I could tell, JAX-B is a required dependency and REST Assured fails fairly fundamentally if the `javax.xml.bind` classes from EE 8 aren't available.

**Comment by johanhaleby** [comment]
I'll look into this and see if I can make any progress.

**Comment by johanhaleby** [comment]
Can you still use Java 8 with the jakarta API?

**Comment by johanhaleby** [comment]
@wilkinsona And do you think it'll be fine to use Jakarta EE 9 for rest assured for the time being? Maybe I should drop support for Java 8 in the future, but I'd like to wait a bit longer if possible.

**Comment by wilkinsona** [comment]
Requiring Jakarta EE 9 at this point is a pretty big jump. We're heading in that direction for Spring but our releases that make the jump aren't due until the end of this year. Even once they're out, we'll continue to support EE 8 for quite some time as we're anticipating lots of users sticking with EE 8. In short – and I know this isn't going to be easy – I think supporting both EE 8 and EE 9 is going to be necessary across the ecosystem for the next 24-36 months.

To minimise that burden, I'd try to isolate yourself as much as possible from Jakarta EE and minimise your dependencies upon its APIs. Where you have to depend on it, try to make those dependencies optional. For example, if a REST API isn't doing anything with XML, perhaps REST Assured could avoid loading any JAX-B classes.

**Comment by lulseged** [comment]
Easier to keep 4.x.x in a separate maintenance branch for JAVA EE 8 and use a different branch(preferably master) for Jakarta 9.x.x.

**Comment by wilkinsona** [comment]
I'm not sure that doubling a project's maintenance burden is going to be easier, particularly when resources may be stretched pretty thin already.

**Comment by johanhaleby** [comment]
@wilkinsona Thank you for your explaination. I think that I'll try to make the dependency to jaxb optional.

**Comment by johanhaleby** [comment]
@wilkinsona Maybe you could clarify some more points and give me some additional tips if you don't mind? :) I'm in the process of making the JAXB dependency optional and I think that'll work. However, I have two questions:

1. Is the "removal" (making it optional) of JAXB what's actually required for Spring Boot to work with REST Assured again?
2. I'd like to add support for Jakarta EE 9 but, if I understand it correctly, Jakarta EE 8 and 9 have the same groupId and artifactId but different version numbers. Currently, in trunk, I have (optional) dependencies on version 2.3.x of `jakarta.xml.bind:jakarta.xml.bind-api` and `org.glassfish.jaxb:jaxb-runtime` (Jakarta EE 8). But I'd like to have version 3.x of these two libraries at the same time as an optional dependency in Maven. Is there a way to work around this that you're aware of?

**Comment by wilkinsona** [comment]
> Is the "removal" (making it optional) of JAXB what's actually required for Spring Boot to work with REST Assured again?

I can't say for certain without being able to try it, but I would expect that for a large percentage of users this will be absolutely fine as they'll be working with JSON-based APIs and won't care about XML

> Is there a way to work around this that you're aware of?

Different projects have been approaching this in different ways:

1. Maintain one set of source code (typically using the `javax.*` APIs) and then use the [Eclipse Transformer](https://github.com/eclipse/transformer) to generate classes that use the `jakarta.*` APIs automatically. Typically published as separate modules and users then need to select the right one. Hibernate has taken this approach.
2. Use an older version of the `javax` APIs alongside the `jakarta` APIs in the same module and take advantage of the different group IDs. For JAX-B that would be `javax.xml.bind:jaxb-api:2.3.1` and `jakarta.xml.bind:jakarta.xml.bind-api:3.0.1`. Thymeleaf has taken this approach.
3. Maintain two separate source modules, one that uses the `javax` APIs and one that uses the `jakarta` APIs. Published as separate modules and users then need to select the right one.

I think 1 works well for a project, like Hibernate, that implements an EE API (JPA in this case). 2 seems to work well for a project that only consumes an EE API and may be a better fit for REST Assured, particularly if you can switch between javax JAX-B and Jakarta JAX-B at runtime depending on what's available on the classpath.

**Comment by johanhaleby** [comment]
@wilkinsona Thanks a lot for the clarification, this is really invaluable to me!

When it comes to question number 1, I was mainly asking from a spring-boot perspective. Because I think Spring Boot has removed rest-assured in version 3 if I'm not mistaken. My goal is to get it back :) But I wonder what I need to do in order to get it back as a managed dependency by spring boot 3? Is it enough to remove the explicit JAXB dependency, or is jakarta ee required, or/and is it something else? If you don't know, do you know who I should talk to?

**Comment by wilkinsona** [comment]
We removed REST Assured support due to the hard dependency on JAX-B from EE 8. Once REST Assured can be used in an EE 9 environment (even if that means there's no XML support), we'll be more than happy to reinstate support for it. The same goes for Spring REST Docs from which REST Assured support has also been temporarily removed for the same reasons.

**Comment by wilkinsona** [comment]
There's plenty of time. While Spring Boot 3.0.0-M1 is scheduled for this Thursday, we're quite comfortable with releasing it without REST Assured support. We know that Jakarta EE 9 support across the ecosystem is going to take a while which is, in part, why we're having a longer development cycle for 3.0.0 (almost 12 months) and it won't GA until November 2022.

The dates for 3.0's milestones are [on GitHub](https://github.com/spring-projects/spring-boot/milestones). They're every two months until M5 in September and then there's an RC in October before GA in November. That's the current plan anyway.

**Comment by johanhaleby** [comment]
@wilkinsona Thanks again for all your feedback and for responding so fast.

Unfortunatley I ran into an issue with Intellij (https://youtrack.jetbrains.com/issue/IDEA-286871) and currently I can't even compile REST Assured so I have to use Sublime which is suboptimal. But regardless of EE 9 or not, I still want to make a new release as soon as I can, it's long overdue.

**Comment by spencergibb** [comment]
Spring Cloud contract has a hard dependency. We release a week after boot, if that could be possible, we would appreciate it

**Comment by johanhaleby** [comment]
@spencergibb Thanks for letting me know! I'll try my very best to release a new version this friday!

**Comment by johanhaleby** [comment]
@spencergibb Or do you mean I should try to create a release sooner (before Thursday?)

**Comment by thetric** [comment]
@johanhaleby I did not encounter this problem with Maven 3.8.4 and this change in the root `pom.xml`:

```xml
    <parent>
        <groupId>org.sonatype.oss</groupId>
        <artifactId>oss-parent</artifactId>
        <version>9</version>
    </parent>
```

I had to upgrade this version because the old one still used JCenter as repository. Now it fails due to a compilation error related to the JAXB migration:

```
[INFO] Reactor Summary for REST Assured Parent POM 4.4.1-SNAPSHOT:
[INFO] 
[INFO] REST Assured Parent POM ............................ SUCCESS [  0.940 s]
[INFO] REST Assured Modules ............................... SUCCESS [  0.024 s]
[INFO] json-schema-validator .............................. SUCCESS [  6.073 s]
[INFO] rest-assured-common ................................ SUCCESS [  6.697 s]
[INFO] json-path .......................................... SUCCESS [  7.419 s]
[INFO] xml-path ........................................... SUCCESS [  7.257 s]
[INFO] REST Assured ....................................... FAILURE [ 10.094 s]
[INFO] spring-commons ..................................... SKIPPED
[INFO] spring-mock-mvc .................................... SKIPPED
[INFO] Scala Support ...................................... SKIPPED
[INFO] spring-web-test-client ............................. SKIPPED
[INFO] kotlin-extensions .................................. SKIPPED
[INFO] spring-mock-mvc-kotlin-extensions .................. SKIPPED
[INFO] rest-assured-all ................................... SKIPPED
[INFO] REST Assured Examples .............................. SKIPPED
[INFO] Scalatra Example ................................... SKIPPED
[INFO] Scalatra REST service example ...................... SKIPPED
[INFO] REST Assured Java integration tests ................ SKIPPED
[INFO] Spring MVC example ................................. SKIPPED
[INFO] Scala Example ...................................... SKIPPED
[INFO] Scala Spring Mock MVC Example ...................... SKIPPED
[INFO] Kotlin Example ..................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  40.829 s
[INFO] Finished at: 2022-01-18T13:33:51+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.gmaven:gmaven-plugin:1.5:compile (default) on project rest-assured: startup failed:
[ERROR] file:/C:/Users/broj/Downloads/rest-assured-master/rest-assured/src/main/groovy/io/restassured/internal/mapping/JaxbMapper.groovy: 43: unable to resolve class javax.xml.bind.JAXBContext
[ERROR]  @ line 43, column 36.
[ERROR]            javax.xml.bind.JAXBContext jaxbContext = factory.create(object.getClass(), charset)
[ERROR]                                       ^
[ERROR] 
[ERROR] file:/C:/Users/broj/Downloads/rest-assured-master/rest-assured/src/main/groovy/io/restassured/internal/mapping/JaxbMapper.groovy: 44: unable to resolve class javax.xml.bind.Marshaller
[ERROR]  @ line 44, column 35.
[ERROR]            javax.xml.bind.Marshaller marshaller = jaxbContext.createMarshaller()
[ERROR]                                      ^
[ERROR] 
[ERROR] 2 errors
```

**Comment by spencergibb** [comment]
Friday is fine

**Comment by johanhaleby** [comment]
@thetric That works for me too! It's not optimal, but at least now I can work in Intellij. Thanks a lot!

**Comment by wilkinsona** [comment]
Thanks very much for this, @johanhaleby. I've reinstated support for REST Assured in Spring REST Docs and Spring Boot. It's available in their 3.0.0-SNAPSHOTs and will be in their 3.0.0-M2 releases in March.

**Comment by prdoyle** [comment]
Just wanted to drop a note to thank you all for your efforts. Spring and REST Assured are both invaluable tools for us, and your efforts are greatly appreciated!
